File: Search/FibonacciSearch/fibonacci_search.cpp
//Search algorithm for sorted list, with complexity of the order O(log(n)).
#include <iostream>
#include <vector>

using namespace std;

int fibsearch(vector<int> arr,int x){
  //indicates first two fibonacci nummbers
  int fib2 = 0,fib1 = 1;
  int fib = fib1 + fib2;

  int len = arr.size();
  //to calculate smallest fibonacci no greater than or equal to len
  while(fib < len){
    fib2 = fib1;
    fib1 = fib;
    fib = fib1 + fib2;
  }
  int leftout = -1; //the leftout list for the array
  while(fib > 1){
    //to get a valid index
    int index = min(leftout + fib2 , len - 1);

    if(arr[index] > x){
      fib = fib2;
      fib1 -= fib2;
      fib2 = fib - fib1;
    }
    else if(arr[index]<x){
      fib = fib1;
      fib1 = fib2;
      fib2 = fib - fib1;
      leftout = index;
    }
    else{
      return index;
    }
  }

  if(fib1 && arr[leftout + 1] == x){
    return leftout+1;
  }
  else{
    return -1;
  }
}
int main(){
  vector<int> arr;
  int n;//no of elements
  cin>>n;
  for(int i=0;i<n;i++){
    int temp;
    cin>>temp;
    arr.push_back(temp);
  }
  int search_no;//element to be searched
  cin>>search_no;
  int ans = fibsearch(arr,search_no);
  if(ans!=-1){
    cout<<"Given element found at index "<<ans<<endl;
  }
  else{
    cout<<"Given element is not found"<<endl;
  }
  return 0;
}

File: Search/FibonacciSearch/fibonacci_search.py
def fibonacci_search(arr, x, n):
    """
    Returns index of x if present, else returns -1
    :param arr:
    :param x:
    :param n:
    :return:
    """
    fib2 = 0   # (m-2)'th Fibonacci No.
    fib1 = 1   # (m-1)'th Fibonacci No.
    fib = fib2 + fib1   # m'th Fibonacci

    while fib < n:
        fib2 = fib1
        fib1 = fib
        fib = fib2 + fib1

    offset = -1
    while fib > 1:
        # to get a valid index
        i = min(offset + fib2, n - 1)

        if arr[i] < x:
            fib = fib1
            fib1 = fib2
            fib2 = fib - fib1
            offset = i

        elif arr[i] > x:
            fib = fib2
            fib1 = fib1 - fib2
            fib2 = fib - fib1

        else:
            return i

    if fib1 and arr[offset + 1] == x:
        return offset + 1

    # element not found. return -1
    return -1


if __name__ == "__main__":
    n = int(input("Enter the number of elements: "))
    print("Enter the elements each line")
    arr = []
    for i in range(n):
        arr.append(int(input()))
    x = int(input("Enter element to be searched: "))
    idx = fibonacci_search(arr, x, n)
    if idx == -1:
        print("Element not found!")
    else:
        print("Element found at position: "+str(idx+1))
